src/main.rs

```diff
@@ -1,76 +1,123 @@
 use egg::*;
 use std::env::*;
 use std::fs::*;
 use std::time::*;
+use std::time::{Duration, Instant};
+use tamago::model::*;
+use tamago::optimize::*;
+use tamago::resnet50;
+use tamago::rewrites::*;
+use tamago::{parse::*, verify::*};
 
 fn main() {
-    //prove_taso_rules();
-    optimize();
-    //convert_rw_rules();
-    //test();
+    // Parse arguments
+    let matches = App::new("Tamago")
+        .arg(
+            Arg::with_name("mode")
+                .short("m")
+                .long("mode")
+                .takes_value(true)
+                .help("Mode to run, can be verify, optimize, test, convert"),
+        )
+        .arg(
+            Arg::with_name("model")
+                .short("d")
+                .long("model")
+                .takes_value(true)
+                .help("Specify a pre-defined model to optimize"),
+        )
+        .arg(
+            Arg::with_name("rules")
+                .short("r")
+                .long("rules")
+                .takes_value(true)
+                .help("Provide a file with rewrite rules"),
+        )
+        .arg(
+            Arg::with_name("model_file")
+                .short("f")
+                .long("model_file")
+                .takes_value(true)
+                .help("Provide a file with the input model"),
+        )
+        .get_matches();
+
+    let run_mode = matches.value_of("mode").unwrap_or("optimize");
+    println!("Running mode is: {}", run_mode);
+
+    match run_mode {
+        "optimize" => optimize(matches),
+        "verify" => prove_taso_rules(matches),
+        "test" => test(matches),
+        "convert" => convert_rw_rules(matches),
+        _ => panic!("Running mode not supported"),
+    }
 }
 
-fn convert_rw_rules() {
+fn convert_rw_rules(matches: clap::ArgMatches) {
     env_logger::init();
-    let file = args().nth(1).expect("Pls supply taso rules file.");
+
+    let file = matches
+        .value_of("rules")
+        .expect("Pls supply taso rules file.");
     let taso_rules = read_to_string(file).expect("Something went wrong reading the file");
 
     let converted = parse_and_convert(&taso_rules);
 
     write("converted.txt", converted).expect("Unable to write file");
 }
 
-
-fn test() {
+fn test(matches: clap::ArgMatches) {
     env_logger::init();
-    let file = args().nth(1).expect("Pls supply example graph file.");
-    let input_graph = read_to_string(file).expect("Something went wrong reading the file");
-    let start = input_graph.parse().unwrap();
 
-    let runner = Runner::<Mdl, TensorAnalysis, ()>::default()
-        .with_expr(&start);
-
-    runner.egraph.dot().to_svg("target/start.svg").unwrap();
+    let start = resnet50::get_resnet50();
 
-    println!("  Nodes: {}", runner.egraph.total_size());
+    let runner_start = Runner::<Mdl, TensorAnalysis, ()>::default().with_expr(&start);
+    println!("Runner complete!");
+    runner_start
+        .egraph
+        .dot()
+        .to_svg("target/start.svg")
```

Comment:
Should generating the dot/svg perhaps be behind a flag? (at least for most graphs I've worked with this can take some time :) ). Could the egraph be serialized so that one could draw it later?

Reply:
Yeah that makes sense! I can add that later

Reply:
Yeah that makes sense, I can add this later